### DISCUSSION TITLE
Mocking (input for) `useActor`?

### DISCUSSION THREAD

**Discussion by karfau** [statelyai/xstate#1609] [Q&A]
I have a (react) component that receives a property `machineRef: Interpreter<...>` which is used to call `const [machineState, machineSend] = useActor(machineRef)` inside the component. 

What would be a best practice to mock the result of `useActor`, so that I'm in control 
of `machineState.*` or can assert on calls to `machineSend`?

What would I need to provide as input for `useActor` to achieve that?

Mocking `useActor` completely might not be what I want because it's also used "deeper" inside the component (for actors inside a machine that is internal to the component), where I don't want to change the behavior. But maybe there is a way to only mock `useActor` for the case that the input is the provided property?

Any thoughts or ideas?

**Comment by davidkpiano** [comment]
Sure, something like this should work:

```js
const fakeActor = {
  send: (event) => {/* ... */},
  subscribe: () => {
    return { unsubscribe: () => {/* ... */} }
  }
};

const Test = ({ actor }) => {
  const [state, send] = useActor(actor);

  // ...
}

// in tests
render(<Test actor={fakeActor} />)
```

**Reply by karfau** [reply]
Thx for the fast reply!
That looks really nice and simple, but how do I control the value of `state(.context...)` in your example?
Would that be something like
```
subscribe: () => {
    return { 
      unsubscribe: () => {/* ... */},
      context: {/* ... */}
    }
  }
```

**Reply by davidkpiano** [reply]
That's up to you to implement. This is just JavaScript.

Here's an idea for a simple counter actor:

```js
const createCounterActor = () => {
  let count = 0;
  let subscriber;

  return {
    send: (event) => {
      // do whatever you want
      count += 1;
      subscriber?.(count);
    },
    subscribe: (fn) => {
      subscriber = fn;

      return { unsubscribe: () => {
        subscriber = undefined;
      } }
    }
  }
};
```

**Reply by karfau** [reply]
In my case the `state` doesn't need to be modified by events, the component just needs to read values from `state.context`.
So if I understand your example correctly the `subscribe` method get's the `state` object as an argument and could write the data to it that the component needs to read?

```
const fakeActor = (context) => {
  return {
    send: (event) => { /* ... */ },
    subscribe: (state) => {
      state.context = context;
      return { unsubscribe: () => {} }
    }
  }
};
```

**Reply by davidkpiano** [reply]
Can you share exactly what you're testing? You can test this directly with `useMachine(...)` instead.

**Reply by karfau** [reply]
I'm not able to provide the complete thing, but I'll be a bit more specific:

I'm trying to create a model based test (using `@xstate/test`) for a non trivial dialog inside an application.

The only properties for the dialog are an `onClose:() => void` handler and (as mentioned above) a `machineRef: Interpreter<...>` which is used for the following things:
- accessing two values from `machineState.context` to initialize  an internal state machine
- using `machineSend` to signal success or failure to the "top level" machine.

Is that enough information? it feels like I already wrote all of this before...
